## Analysis for Figures 1A-C
**All of the code in this page is meant to be run on the command line unless otherwise stated.**
These are the commands we used to generate sample/sequence rarefaction curves and calculate different alpha diversity metrics:
```{r eval=FALSE, engine='bash'}
# Rename taxonomy
sed -i 's/Consensus.Lineage/taxonomy/g' OTU-Table-16S.tsv 

# Convert table to biom format
biom convert -i OTU-Table-16S.tsv -o OTU-Table-16S.biom --table-type="OTU table" --to-json --process-obs-metadata taxonomy

# Download phylogenetic tree 
wget http://downloads.hmpdacc.org/data/HMQCP/rep_set_v35.tre.gz

# Get OTU ids present in the analyzed samples only
tail -n +2 OTU-Table-16S.tsv | cut -f1 > OTUs_to_keep.txt

# Prune the phylogenetic tree 
filter_tree.py -i rep_set_v35.tre -t OTUs_to_keep.txt -o pruned_rep_set_v35.tree

# Filter low abundant otus
filter_otus_from_otu_table.py -i OTU-Table-16S.biom -n 2 -s 3 -o OTU-Table-16S.filtered.biom

# Get the stats for the filtered table
biom summarize-table -i OTU-Table-16S.filtered.biom -o OTU-Table-16S.filtered.summary

# Rarefy OTU table
multiple_rarefactions_even_depth.py -i OTU-Table-16S.filtered.biom -d 1802 -o Rarefied_Tables/

# Calculate alpha diversity on each of the rarefied tables
alpha_diversity.py -i Rarefied_Tables/ -m observed_species,PD_whole_tree,shannon,equitability -t pruned_rep_set_v35.tree -o Alpha_diversity/

# Collate different alpha diversity tables
collate_alpha.py -i Alpha_diversity/ -o Alpha_Diversity_Collated/

# Parametric testing of alpha diversity
for i in Alpha_Diversity_Collated/*; 
do compare_alpha_diversity.py -i $i -m Qiime.map -c Description -o $i.statistics; 
done

for i in Alpha_Diversity_Collated/*.statistics; 
do measure=$(echo $i | cut -f2 -d"/" | cut -f1 -d"."); 
echo $measure; grep -A 1 "p-value" $i/Description_stats.txt; 
done

# Rarefaction
```

